src/swell/utilities/jinja2.py

```diff
@@ -7,28 +7,75 @@
 # --------------------------------------------------------------------------------------------------
 
 
-import jinja2
+import jinja2 as j2
 
 
 # --------------------------------------------------------------------------------------------------
 
 
-def template_string_jinja2(logger, templated_string, dictionary_of_templates):
+class SilentUndefined(j2.Undefined):
+    """
+    A custom undefined class that doesn't raise errors when variables are missing and returns the
+    original template variable placeholder.
+    """
+    def __getattr__(self, name):
+        # Return a new SilentUndefined instance but append the attribute access to the name.
+        return SilentUndefined(name=f"{self._undefined_name}.{name}")
 
-    # Load the templated string
-    t = jinja2.Template(templated_string, trim_blocks=True, lstrip_blocks=True,
-                        undefined=jinja2.StrictUndefined)
+    def __getitem__(self, key):
+        # Similar to __getattr__, return a new instance with the key access incorporated.
+        if isinstance(key, str):
+            return SilentUndefined(name=f"{self._undefined_name}['{key}']")
+        return SilentUndefined(name=f"{self._undefined_name}[{key}]")
 
-    # Render the templates using the dictionary
-    string_rendered = t.render(dictionary_of_templates)
+    def items(self):
+        # Return an empty list when items method is called.
+        return []
 
-    logger.assert_abort('{{' not in string_rendered, f'In template_string_jinja2 ' +
-                        f'the output string still contains template directives. ' +
-                        f'{string_rendered}')
+    def __str__(self):
+        # Ensure the name returned reflects the original template placeholder.
+        return f"{{{{ {self._undefined_name} }}}}"
 
-    logger.assert_abort('}}' not in string_rendered, f'In template_string_jinja2 ' +
-                        f'the output string still contains template directives. ' +
-                        f'{string_rendered}')
+    def __repr__(self):
+        return str(self)
+
+
+# --------------------------------------------------------------------------------------------------
+
+
+def template_string_jinja2(logger, templated_string, dictionary_of_templates,
+                           allow_unresolved=False):
+
+    # Handling of templates that cannot be resolved
+    # ---------------------------------------------
+    undefined = SilentUndefined if allow_unresolved else j2.StrictUndefined
+
+    # Create the Jinja2 environment
+    # -----------------------------
+    env = j2.Environment(undefined=undefined)
+
+    # Load the algorithm template
+    # ---------------------------
+    template = env.from_string(templated_string)
+
+    # Render the template hierarchy
+    # -----------------------------
+    try:
+        string_rendered = template.render(dictionary_of_templates)
+    except j2.exceptions.UndefinedError as e:
```

Comment:
Since we're just re-throwing the exception, any reason not to just do this?

```suggestion
    except Exception as e:
```

Otherwise, the Python linter complains (correctly) that `string_rendered` is potentially undefined below.

Or, is the point that we do actually want to silently ignore all exceptions here _except_ the `UndefinedError` exception? If so, suggest a comment as to why.

Reply:
It's the latter. There is a description under `ask_questions_and_configure_suite` method in  `prepare_config_and_suite.py`. Essentially, with the way it is designed now, there are two jinja passes one of which being non-template resolving (non-exhaustive) vs. template resolving. Does this make sense?

For instance, I had to add `items(self)` in the `SilentUndefined` class to make sure jinja2 ignores the below part:

```
        {%- for key, value in scheduling["RunJediVariationalExecutable"]["directives"][model_component].items() %}
```

Reply:
I will add a comment though.